### PR TITLE
fix(gc): do not garbage collect block info past gc_stop_height

### PR DESCRIPTION
**Pull request by bowenwang1996**
When we do state sync, we clear all data before the sync block. However, this means block infos that are needed to compute gc_stop_height afterwards may also be cleared, which will cause failures when we do gc after state sync.

Test plan
---------
`test_gc_after_state_sync` passes

### CODE REVIEW

**Comment by bowenwang1996** [comment]
@Kouprin please let me know if some invariant is broken with this change.

**Comment by bowenwang1996** [comment]
@Kouprin No. We run state sync when we are more than one epoch behind the head. Let's consider the following situation. A node was running normally in the network but for some reason it shut down (maybe out of disk space for example). Later it was revived and at that point, it was just more than one epoch behind the head of the chain and therefore state sync is initiated.

**Comment by Kouprin** [comment]
What you're saying means we GCed old data because we can't use it anymore but we want to keep BlockInfos to restore `gc_stop_height` to clear the data. This is weird.

I see the following solutions:
1. We need to reconsider which data is useful and why and update `reset_data_pre_state_sync` to keep all data that is relatively new, not only BlockInfos.
2. We can update `gc_stop_height` calculation is such way that if we can't traverse 5 epochs then gc_stop_height == genesis height.

I don't see why we need p. 1 and we can go with p. 2 easily.

**Comment by bowenwang1996** [comment]
> We need to reconsider which data is useful and why and update reset_data_pre_state_sync to keep all data that is relatively new, not only BlockInfos.

I have considered this as well. I initially wanted to do it in such a way that reset_data_pre_state_sync doesn't go beyond `gc_stop_height`, but then I realize that if we clear the entire state, this will cause issues such as having blocks without the corresponding state to apply them.

> We can update gc_stop_height calculation is such way that if we can't traverse 5 epochs then gc_stop_height == genesis height.

I like the idea :)

File: chain/chain/src/chain.rs [hunk 1]
```diff
@@ -551,7 +551,13 @@ impl Chain {
     ) -> Result<(), Error> {
         let head = self.store.head()?;
         let tail = self.store.tail()?;
-        let gc_stop_height = self.runtime_adapter.get_gc_stop_height(&head.last_block_hash)?;
+        let gc_stop_height = match self.runtime_adapter.get_gc_stop_height(&head.last_block_hash) {
+            Ok(height) => height,
+            Err(e) => match e.kind() {
+                ErrorKind::DBNotFoundErr(_) => self.genesis.height(),
```
Comment (Kouprin) [inline chain/chain/src/chain.rs:557]: `return Ok(())` because we can't go forward. Please add a comment describing what's happening here.

**Comment by Kouprin** [comment]
You can also add the following python test:
- create cluster with 2 nodes
- turn off 2nd node at 2nd epoch
- turn on the node at 4th epoch
- make sure that after 10th epoch the sets of available blocks are the same

**Comment by bowenwang1996** [comment]
@Kouprin 

> It would be awesome not to forget about block_info_block_exists.

That check isn't right because during header sync we can have block info without having block. The correct statement is that there is an injection from block info to block headers. I've fixed the check.

**Pull request merged**